plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1868,14 +1868,15 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
    _pending_block_deadline = block_timing_util::calculate_producing_block_deadline(_produce_block_cpu_effort, block_time);
    if (in_speculating_mode()) { // if we are producing, then produce block even if deadline has passed
       // speculative block, no reason to start a block that will immediately be re-started, set deadline in the future
-      // a block should come in during this time, if not then just keep creating the block every produce_block_cpu_effort
-      if (now + fc::milliseconds(config::block_interval_ms/10) > _pending_block_deadline) {
-         _pending_block_deadline = now + _produce_block_cpu_effort;
+      // a block should come in during this time, if not then just keep creating the block every block_interval_ms
```

Comment:
that three line comment doesn't seem very clear to me.

Reply:
I re-read the comments. Not sure what I can change/add there. Any ideas?

Reply:
I guess I don't understand why `block_interval_ms` is the right deadline, why we thought that 50ms was better before, etc...
I know we talked about it during the meeting but I don't really understand the reasoning behind it. Maybe you can explain it to me in a conversation and then I can suggest a new comment.

Reply:
This is for speculative block production. Before we used `_produce_block_cpu_effort` which is `block_interval_ms - block-offset`. The block offset is just a configurable amount that provides time for the block to reach the next block producer. I don't really see a point of factoring that into this. The 50ms before was just a threshold for if the `_produce_block_cpu_effort` should be added on to now. We used to wait indefinitely for a block to come in when creating a speculative block. We decided it was good to re-start a speculative block every block interval so that any speculative transactions would be able to run using the current block time. This change simplified the logic and just uses `block_interval_ms` so we get a new speculative block every 500ms. In general we expect a new block to be received in that time frame which will abort the speculative block, apply the received block, and start a new speculative block.

Reply:
Wow, this is a great explanation. Thank you Kevin.

I tried to translate my understanding into a comment:

```
// in speculating mode, we want to start creating blocks on a regular basis so that we can
// execute speculative transaction with the current block time.
// If a block is received, we will be able to abort the speculative block when the current
// transaction completes (a maximum of 150ms) - this because we currently don't interrupt transactions.
// We could set a `_pending_block_deadline` less than 150ms in order to minimize the delay before
// an incoming block can be processed. However the drawback is that the speculative block could
// potentially be restarted multiple times before incoming blocks.
```

PS: Why is it a problem if the speculative  block is restarted?

Reply:
There is no problem to restart the speculative block, just wasteful. Wasteful in the time it takes to abort the current block and start a new one; but mostly wasteful in that you abort the trx at the deadline and it has to be restarted on the next speculative block.

Reply:
> There is no problem to restart the speculative block, just wasteful. Wasteful in the time it takes to abort the current block and start a new one; but mostly wasteful in that you abort the trx at the deadline and it has to be restarted on the next speculative block.

Do we care if it is wasteful? Since we didn't get a block we don't have much to do anyways, right? And maybe that makes nodeos more responsive to things like `get_info`?

Reply:
There could be thousands of trxs queued up ready to execute. I think that is the important thing to optimize in this case.

This is one case where knowing if the node is a relay node vs an API node would allow us to make different choices.

Reply:
Thanks for the comment update!